File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        
        # Initialize arrays with default values
        left = [1] * n
        right = [1] * n
        ans = [0] * n
        
        # Fill the left product array
        for i in range(1, n):
            left[i] = left[i - 1] * nums[i - 1]
        
        # Fill the right product array
        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]
        
        # Calculate the result using left and right products
        for i in range(n):
            ans[i] = left[i] * right[i]
        
        return ans

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n=nums.length;

        //APP-1 =With division

        // int productWithoutZero=1;
        // int count=0;
        // for(int i=0;i<nums.length;i++){
        //     if(nums[i]==0){
        //         count++;
        //     }else{
        //         productWithoutZero*=nums[i];
        //     }
        // }
        // int result[]=new int[nums.length];
        // for(int i=0;i<nums.length;i++){
        //     if(nums[i]!=0){
        //         if(count>0){
        //             result[i]=0;
        //         }
        //         else{
        //             result[i]=productWithoutZero/nums[i];
        //         }
        //     }
        //     else{
        //         if(count>1){
        //             result[i]=0;
        //         }
        //         else{
        //             result[i]=productWithoutZero;
        //         }
        //     }
        // }
        // return result;


        //Approach 2:without division but use extra spaces
        
        int left[]=new int[n];
        left[0]=1;
        int right[]=new int[n];
        right[n-1]=1;
        for(int i=1;i<n;i++){
            left[i]=left[i-1]*nums[i-1];
        }
        for(int i=n-2;i>=0;i--){
            right[i]=right[i+1]*nums[i+1];
        }
        int result[]=new int[n];
        for(int i=0;i<n;i++){
            result[i]=left[i]*right[i];
        }
        return result;
    }
}